### DISCUSSION TITLE
A little high CPU usage

### DISCUSSION THREAD

**Discussion by tesuji** [yshui/picom#203] [General]
### Platform

```
Ubuntu 18.04.2 LTS (Bionic Beaver) 4.18.0-25-generic #26~18.04.1-Ubuntu SMP Thu Jun 27 07:28:31 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

### GPU, drivers, and screen setup

- Intel(R) Core(TM) i7-4800MQ CPU @ 2.70GHz Integrated graphics, one monitor
- `xf86-video-*` version: None
- `mesa-utils` version: 8.4.0-1

```console
% glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) Haswell Mobile  (0x416)
    Version: 19.0.2
    Accelerated: yes
    Video memory: 1536MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Haswell Mobile 
OpenGL core profile version string: 4.5 (Core Profile) Mesa 19.0.2
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.0 Mesa 19.0.2
OpenGL shading language version string: 1.30
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.1 Mesa 19.0.2
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
```

### Environment

xfce4 v4.12.4

### Compton version

- `compton --version`: v7, commit d45ed14

```console
% compton --diagnostics
[ 07/19/2019 21:53:41.184 parse_config_libconfig WARN ] vsync option will take a boolean from now on. "opengl-swc" is interpreted as "true" for compatibility, but this will stop working soon
[ 07/19/2019 21:53:41.187 register_cm FATAL ERROR ] Another composite manager is already running
```

### Compton configuration:

<details>
	<summary>compton.conf</summary>

```bash
# ~/compton/.config/compton.conf -- Configuration of compton - A compositor for X11
#
# Use `kill -USR1 $(pidof compton)` for compton to reload this file.

# -- Backend ---------------------------------------------------------------------

# Backend to use: "xrender" or "glx".
# GLX backend is typically much faster but depends on a sane driver.
# Change to xrender if issues when using AMD's Catalyst driver
backend = "glx";

# -- GLX backend -----------------------------------------------------------------

# GLX backend: Avoid using stencil buffer, useful if you don’t have a
# stencil buffer. Might cause incorrect opacity when rendering transparent
# content (but never practically happened) and may not work with
# --blur-background. Tests show a 15% performance boost. Recommended.
glx-no-stencil = true;

# GLX backend: Copy unmodified regions from front buffer instead of redrawing them all.
# Tests with nvidia-drivers show a 10% decrease in performance when the whole
# screen is modified, but a 20% increase when only 1/4 is.
# Tests on nouveau show terrible slowdown.
# Useful with --glx-swap-method, as well.
#
# DEPRECATION: The developer who added this flag in the first place has
# said "do not use [this flag]" chjj#168 (comment).
#glx-copy-from-front = false;

# GLX backend: Use MESA_copy_sub_buffer to do partial screen update.
# Tests on nouveau shows a 200% performance boost when only 1/4 of the screen is updated.
# May break VSync and is not available on some drivers.
# Overrides --glx-copy-from-front.
#
# DEPRECATION: https://github.com/yshui/compton/releases/tag/v3
#glx-use-copysubbuffermesa = true;

# GLX backend: Avoid rebinding pixmap on window damage.
# Probably could improve performance on rapid window content changes, but is known
# to break things on some drivers (LLVMpipe).
# Recommended if it works.
#glx-no-rebind-pixmap = true;

# -- Shadow ----------------------------------------------------------------------

shadow = false; # Enabled client-side shadows on windows.
shadow-radius = 5; # The blur radius for shadows. (default 12)
shadow-offset-x = -5; # The left offset for shadows. (default -15)
shadow-offset-y = -5; # The top offset for shadows. (default -15)
shadow-opacity = 1; # The translucency for shadows. (default .75)

# Set if you want different colour shadows
#shadow-red = 0.0;
#shadow-green = 0.0;
#shadow-blue = 0.0;

# The shadow exclude options are helpful if you have shadows enabled. Due to
# the way compton draws its shadows, certain applications will have visual glitches
# (most applications are fine, only apps that do weird things with xshapes or argb are affected).
# This list includes all the affected apps I found in my testing. The "! name~=''"
# part excludes shadows on any "Unknown" windows, this prevents a visual glitch with
# the XFWM alt tab switcher.
shadow-exclude = [
 "n:e:Notification",
 "n:e:Docky",
 "g:e:Synapse",
 "g:e:Conky",
 "n:w:*Firefox*",
 "n:w:*Chromium*",
 "n:w:*dockbarx*",
 "class_g ?= 'Cairo-dock'",
 "class_g ?= 'Xfce4-notifyd'",
 "class_g ?= 'Xfce4-power-manager'",
 "class_g ?= 'Notify-osd'",
 "_GTK_FRAME_EXTENTS@:c"
];

# Avoid drawing shadow on all shaped windows (see also: --detect-rounded-corners)
shadow-ignore-shaped = false;

# -- Opacity ---------------------------------------------------------------------

inactive-opacity = 1;
active-opacity = 1;
frame-opacity = 1;
inactive-opacity-override = false;

# Dim inactive windows. (0.0 - 1.0)
#inactive-dim = 0.2;
# Do not let dimness adjust based on window opacity.
#inactive-dim-fixed = true;
# Blur background of transparent windows. Bad performance with X Render backend.
# GLX backend is preferred.
#blur-background = true;
# Blur background of opaque windows with transparent frames as well.
#blur-background-frame = true;
# Do not let blur radius adjust based on window opacity.
blur-background-fixed = false;
blur-background-exclude = [
    "window_type = 'dock'",
    "window_type = 'desktop'"
];

# https://wiki.archlinux.org/index.php/Compton#Tabbed_windows_(shadows_and_transparency)
opacity-rule = [
  "95:class_g = 'URxvt' && !_NET_WM_STATE@:32a",
  "0:_NET_WM_STATE@[0]:32a *= '_NET_WM_STATE_HIDDEN'",
  "0:_NET_WM_STATE@[1]:32a *= '_NET_WM_STATE_HIDDEN'",
  "0:_NET_WM_STATE@[2]:32a *= '_NET_WM_STATE_HIDDEN'",
  "0:_NET_WM_STATE@[3]:32a *= '_NET_WM_STATE_HIDDEN'",
  "0:_NET_WM_STATE@[4]:32a *= '_NET_WM_STATE_HIDDEN'"
];

# -- Fading ----------------------------------------------------------------------

# Fade windows during opacity changes.
fading = false;
# The time between steps in a fade in milliseconds. (default 10).
fade-delta = 4;
# Opacity change between steps while fading in. (default 0.028).
fade-in-step = 0.03;
# Opacity change between steps while fading out. (default 0.03).
fade-out-step = 0.03;
# Fade windows in/out when opening/closing
#no-fading-openclose = true;

# Specify a list of conditions of windows that should not be faded.
fade-exclude = [ ];

# -- Other -----------------------------------------------------------------------

# Try to detect WM windows and mark them as active.
mark-wmwin-focused = true;
# Mark all non-WM but override-redirect windows active (e.g. menus).
mark-ovredir-focused = true;
# Use EWMH _NET_WM_ACTIVE_WINDOW to determine which window is focused instead of
# using FocusIn/Out events. Usually more reliable but depends on a EWMH-compliant WM.
use-ewmh-active-win = true;
# Detect rounded corners and treat them as rectangular when --shadow-ignore-shaped is on.
detect-rounded-corners = true;

# Detect _NET_WM_OPACITY on client windows, useful for window managers not
# passing _NET_WM_OPACITY of client windows to frame windows.
# This prevents opacity being ignored for some apps.
# For example without this enabled my xfce4-notifyd is 100% opacity no matter what.
detect-client-opacity = true;

# Specify refresh rate of the screen.
# If not specified or 0, compton will try detecting this with X RandR extension.
refresh-rate = 0;

# Set VSync method. VSync methods currently available:
# none: No VSync
# drm: VSync with DRM_IOCTL_WAIT_VBLANK. May only work on some drivers.
# opengl: Try to VSync with SGI_video_sync OpenGL extension. Only work on some drivers.
# opengl-oml: Try to VSync with OML_sync_control OpenGL extension. Only work on some drivers.
# opengl-swc: Try to VSync with SGI_swap_control OpenGL extension.
#             Only work on some drivers. Works only with GLX backend.
#             Known to be most effective on many drivers. Does not actually
#             control paint timing, only buffer swap is affected, so it doesn’t
#             have the effect of --sw-opti unlike other methods. Experimental.
# opengl-mswc: Try to VSync with MESA_swap_control OpenGL extension. Basically
# the same as opengl-swc above, except the extension we use.
# (Note some VSync methods may not be enabled at compile time.)
vsync = "opengl-swc";

# Enable DBE painting mode, intended to use with VSync to (hopefully) eliminate tearing.
# Reported to have no effect, though.
dbe = false;

# Limit compton to repaint at most once every 1 / refresh_rate second to boost performance.
# This should not be used with --vsync drm/opengl/opengl-oml as they essentially does
# --sw-opti's job already, unless you wish to specify a lower refresh rate than the actual value.
sw-opti = false;

# Unredirect all windows if a full-screen opaque window is detected, to maximize
# performance for full-screen windows, like games.
# Known to cause flickering when redirecting/unredirecting windows.
# paint-on-overlay may make the flickering less obvious.
# Fix fullscreen tearing by setting to `false'
unredir-if-possible = false;

# Specify a list of conditions of windows that should always be considered focused.
focus-exclude = [ ];

# Use WM_TRANSIENT_FOR to group windows, and consider windows in the same group focused at the same time.
detect-transient = true;
# Use WM_CLIENT_LEADER to group windows, and consider windows in the same group focused at the same time.
# WM_TRANSIENT_FOR has higher priority if --detect-transient is enabled, too.
detect-client-leader = true;

# -- Window type settings --------------------------------------------------------

wintypes:
{
    tooltip =
    {
        # fade: Fade the particular type of windows.
        fade = false;
        # shadow: Give those windows shadow
        shadow = false;
        # opacity: Default opacity for the type of windows.
        opacity = 1;
        # focus: Whether to always consider windows of this type focused.
        focus = true;
    };
    unknown = { fade = false; shadow = false; opacity = 1; focus = true; };
    desktop = { fade = false; shadow = false; opacity = 1; focus = true; };
    dock = { fade = false; shadow = false; opacity = 1; focus = true; };
    toolbar = { fade = false; shadow = false; opacity = 1; focus = true; };
    menu = { fade = false; shadow = false; opacity = 1; focus = true; };
    utility = { fade = false; shadow = false; opacity = 1; focus = true; };
    splash = { fade = false; shadow = false; opacity = 1; focus = true; };
    dialog = { fade = false; shadow = false; opacity = 1; focus = false; };
    normal = { fade = false; shadow = false; opacity = 1; focus = false; };
    dropdown_menu = { fade = false; shadow = false; opacity = 1; focus = true; };
    popup_menu = { fade = false; shadow = false; opacity = 1; focus = true; };
    notify = { fade = false; shadow = false; opacity = 1; focus = true; };
    combo = { fade = false; shadow = false; opacity = 1; focus = true; };
    dnd = { fade = false; shadow = false; opacity = 1; focus = true; };
};
```
</details>

### Steps of reproduction

- Enable compton on xfce
- Open `htop`

### Expected behavior

I think CPU usage should be lower than that. In the older version (v5 IIRC), the CPU in most times is zero.

### Current Behavior

On htop window, I could see that CPU usage is at about 5% +- 2%.

**Comment by tesuji** [comment]
I ran back to v5. The CPU usage is fine again.

**Comment by yshui** [comment]
@lzutao Did you build compton with `-Dbuildtype=release`? 

`compton` uses negligible CPU here, so I don't know what to look at. Can you do a `perf`?

**Comment by tesuji** [comment]
Thanks for the fast reply.

> Did you build compton with `-Dbuildtype=release`?

Yes, I did.

> Can you do a `perf`?

I will.

**Comment by tesuji** [comment]
Here is the [perf-data.zip](https://github.com/yshui/compton/files/3432229/perf-data.zip) that I run with `sudo perf record compton` for `v7`.
You could use `perf report` to view the data.

**Comment by yshui** [comment]
@lzutao can you also attach the compton executable as well?

**Comment by tesuji** [comment]
Here is it. GitHub doesn't support xz file so you might want to unzip first.
[compton_v7.tar.xz.zip](https://github.com/yshui/compton/files/3433794/compton_v7.tar.xz.zip)

**Comment by yshui** [comment]
Hmm, I don't understand why paint_preprocess is taking that much time.

Do you have a lot of windows open?

**Comment by tesuji** [comment]
I only had 5 windows opened.

**Comment by yshui** [comment]
What does `xwininfo -root -children` say?

**Comment by tesuji** [comment]
<details>
<summary>xwininfo -root -children</summary>

```bash
% xwininfo -root -children

xwininfo: Window id: 0x114 (the root window) (has no name)

  Root window id: 0x114 (the root window) (has no name)
  Parent window id: 0x0 (none)
     95 children:
     0x3400047 "Firefox": ("Popup" "Firefox")  1920x371+0+85  +0+85
     0x34000df "Ubuntu Manpage: clang-format - manual page for clang-format 8": ("Popup" "Firefox")  362x33+1557+45  +1557+45
     0x3400234 "Firefox": ("Popup" "Firefox")  320x24+1256+148  +1256+148
     0x3400826 "Firefox": ("Popup" "Firefox")  386x1080+1204+0  +1204+0
     0x3400822 "Firefox": ("Popup" "Firefox")  250x715+1585+76  +1585+76
     0x340023e "Firefox": ("Popup" "Firefox")  248x104+473+415  +473+415
     0x34000e3 "goto source code": ("Popup" "Firefox")  108x33+1186+234  +1186+234
     0x340069e "Firefox": ("Popup" "Firefox")  227x50+0+88  +0+88
     0x4a06d60 "Sublime Text": ("sublime_text" "Sublime_text")  209x423+579+130  +579+130
     0x4a06b85 "Sublime Text": ("sublime_text" "Sublime_text")  209x447+439+609  +439+609
     0x320285d "Xfce Terminal": ("xfce4-terminal" "Xfce4-terminal")  153x296+69+721  +69+721
     0x3a00001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x3400364 "Firefox": ("Popup" "Firefox")  168x304+129+33  +129+33
     0x4a0027b "Sublime Text": ()  10x10+-100+-100  +-100+-100
     0x4a00009 (has no name): ()  1x1+-1+-1  +-1+-1
     0x4a00001 "Sublime Text": ("sublime_text" "Sublime_text")  10x10+10+10  +10+10
     0x2200022 "Whisker Menu": ("wrapper-2.0" "Wrapper-2.0")  400x500+0+550  +0+550
     0x4800001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x4600001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x4400001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x4200001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x3400372 "Firefox": ("Popup" "Firefox")  386x128+1149+952  +1149+952
     0x3400368 "Firefox": ("Popup" "Firefox")  386x979+1534+101  +1534+101
     0x3800001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x3c00001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x3400316 "Firefox": ("Popup" "Firefox")  278x634+1642+77  +1642+77
     0x3400308 "Firefox": ("Popup" "Firefox")  386x778+243+104  +243+104
     0x340030c "Firefox": ("Popup" "Firefox")  386x178+628+230  +628+230
     0x340028f "Firefox": ("Popup" "Firefox")  200x200+0+0  +0+0
     0x140009d "xfdesktop": ()  10x10+-100+-100  +-100+-100
     0x34001bc "Firefox": ("firefox" "Firefox")  1x1+-99+-99  +-99+-99
     0xe0015e "Thunar": ("Thunar" "Thunar")  1x1+-100+-100  +-100+-100
     0xe00149 "Thunar": ("Thunar" "Thunar")  1x1+-100+-100  +-100+-100
     0xe00063 (has no name): ()  1x1+-1+-1  +-1+-1
     0xe00029 "Thunar": ()  10x10+-100+-100  +-100+-100
     0x3400146 "Firefox": ("Popup" "Firefox")  200x200+0+0  +0+0
     0x3600001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x340005b "Firefox": ("Popup" "Firefox")  200x200+0+0  +0+0
     0x10001c1 "xfce4-panel": ("xfce4-panel" "Xfce4-panel")  1x1+-100+-100  +-100+-100
     0x10001be "xfce4-panel": ("xfce4-panel" "Xfce4-panel")  1x1+-100+-100  +-100+-100
     0x1000185 "xfce4-panel": ("xfce4-panel" "Xfce4-panel")  1x1+-100+-100  +-100+-100
     0x3400017 (has no name): ("Toplevel" "Firefox")  100x100+0+0  +0+0
     0x3400014 "Firefox": ()  10x10+-100+-100  +-100+-100
     0x2e00001 "/usr/lib/firefox/firefox": ("/usr/lib/firefox/firefox" "/usr/lib/firefox/firefox")  10x10+10+10  +10+10
     0x3400011 "Firefox": ()  10x10+-100+-100  +-100+-100
     0x3400001 "Firefox": ("firefox" "Firefox")  10x10+10+10  +10+10
     0x3200241 "Xfce Terminal": ()  10x10+-100+-100  +-100+-100
     0x260001a "wrapper-2.0": ("wrapper-2.0" "Wrapper-2.0")  258x472+1575+584  +1575+584
     0x3200009 (has no name): ()  1x1+-1+-1  +-1+-1
     0x3200001 "Xfce Terminal": ("xfce4-terminal" "Xfce4-terminal")  10x10+10+10  +10+10
     0x1400026 (has no name): ()  1x1+-1+-1  +-1+-1
     0x2800008 (has no name): ()  1x1+-1+-1  +-1+-1
     0x1600004 (has no name): ()  10x10+0+0  +0+0
     0x2c00008 (has no name): ()  1x1+-1+-1  +-1+-1
     0x220000d (has no name): ()  1x1+-1+-1  +-1+-1
     0x2600008 (has no name): ()  1x1+-1+-1  +-1+-1
     0x240004e (has no name): ()  1x1+-1+-1  +-1+-1
     0x1000038 (has no name): ()  1x1+-1+-1  +-1+-1
     0x2c00001 "wrapper-2.0": ("wrapper-2.0" "Wrapper-2.0")  10x10+10+10  +10+10
     0x2800001 "wrapper-2.0": ("wrapper-2.0" "Wrapper-2.0")  10x10+10+10  +10+10
     0x2600001 "wrapper-2.0": ("wrapper-2.0" "Wrapper-2.0")  10x10+10+10  +10+10
     0x2400026 "wrapper-1.0": ()  10x10+-100+-100  +-100+-100
     0x2400001 "wrapper-1.0": ("wrapper-1.0" "Wrapper-1.0")  10x10+10+10  +10+10
     0x1800003 (has no name): ()  1x1+-1+-1  +-1+-1
     0x2200001 "wrapper-2.0": ("wrapper-2.0" "Wrapper-2.0")  10x10+10+10  +10+10
     0xc002c2 (has no name): ()  1x1+-1+-1  +-1+-1
     0xc002be (has no name): ()  1920x1049+0+0  +0+0
     0x1000031 (has no name): ()  10x10+0+0  +0+0
     0x2000002 "light-locker": ()  10x10+-100+-100  +-100+-100
     0x1200002 "xcompmgr": ("compton" "xcompmgr")  1x1+0+0  +0+0
     0x1800001 "xfce4-notifyd": ("xfce4-notifyd" "Xfce4-notifyd")  10x10+10+10  +10+10
     0x2000001 "light-locker": ("light-locker" "Light-locker")  10x10+10+10  +10+10
     0x1e00001 "polkit-gnome-authentication-agent-1": ("polkit-gnome-authentication-agent-1" "Polkit-gnome-authentication-agent-1")  10x10+10+10  +10+10
     0x1c00001 "xfce4-power-manager": ("xfce4-power-manager" "Xfce4-power-manager")  10x10+10+10  +10+10
     0xc000ae "Xfwm4": ("xfwm4" "Xfwm4")  5x5+-1000+-1000  +-1000+-1000
     0xc000ad (has no name): ()  10x10+-100+-100  +-100+-100
     0x1600003 (has no name): ()  1x1+-1+-1  +-1+-1
     0x1600001 "xfsettingsd": ("xfsettingsd" "Xfsettingsd")  10x10+10+10  +10+10
     0x1400001 "xfdesktop": ("xfdesktop" "Xfdesktop")  10x10+10+10  +10+10
     0x1000001 "xfce4-panel": ("xfce4-panel" "Xfce4-panel")  10x10+10+10  +10+10
     0xe00001 "Thunar": ("Thunar" "Thunar")  10x10+10+10  +10+10
     0xc00001 "xfwm4": ("xfwm4" "Xfwm4")  10x10+10+10  +10+10
     0xa00003 (has no name): ()  1x1+1+1  +1+1
     0xa00001 "xfce4-session": ("xfce4-session" "Xfce4-session")  10x10+10+10  +10+10
     0x80000c (has no name): ()  1x1+-1+-1  +-1+-1
     0x800001 "NetworkManager Applet": ("nm-applet" "Nm-applet")  10x10+10+10  +10+10
     0xc00188 (has no name): ()  1x1080+-1+0  +-1+0
     0xc00189 (has no name): ()  1x1080+1920+0  +1920+0
     0xc0018a (has no name): ()  1920x1+0+-1  +0+-1
     0xc0018b (has no name): ()  1920x1+0+1080  +0+1080
     0xc00293 (has no name): ()  1920x31+0+1049  +0+1049
     0xc00319 (has no name): ()  1920x1049+0+0  +0+0
     0xc00555 (has no name): ()  1920x1049+0+0  +0+0
     0xc02d37 (has no name): ()  1920x1049+0+0  +0+0
     0xc002c7 (has no name): ()  1920x1080+0+0  +0+0
```

</details>

**Comment by Hespian** [comment]
Hi, is there an update to this? I have the same issue with the latest version on master and it's fine on v5.1

**Comment by yshui** [comment]
@Hespian Can you do a `perf` as well? What is the refresh rate of your monitor?

**Comment by Hespian** [comment]
Here is my perf file: [perf.zip](https://github.com/yshui/picom/files/4529776/perf.zip)
This time I was only able to reproduce the high cpu usage when I switched focus between windows very often, which is not what I observed before. So right now, it's actually not a problem.
I have two screens both running at 60 Hz

Edit: Actually, here is one where cpu usage goes up by just scrolling on a website in chrome. [perf2.zip](https://github.com/yshui/picom/files/4529849/perf2.zip)

**Comment by yshui** [comment]
@Hespian does v5.1's cpu usage not go up if you scroll in chrome?

**Comment by Hespian** [comment]
Actually it does. But not quite as high. So it seems like I can't reproduce the issues I had earlier.

**Comment by yshui** [comment]
> But not quite as high.

What are the numbers?

**Comment by Hespian** [comment]
> What are the numbers?

v5.1: 3-5%
v8: 5-7%

**Comment by yshui** [comment]
Hmm, that doesn't seem very conclusive. And I also am not able to reproduce this problem myself.

I will probably put this on back burner. You are welcomed to investigate on your own. If you need any help, just ask.

**Comment by quid256** [comment]
Hi, I have the same issue, I'm on Ubuntu as well. I've found that it only occurs with the the glx backend -- the xrender backend consistently uses <1% CPU, but the old glx and experimental glx backends both use >7% consistently. I also noticed that with glx on and vsync off, picom starts using upwards of 20%.

I ran a git bisect. On my machine, the glx backend performance gets significantly worse (CPU usage around 30%) at e80ff85 and stays that bad until 8c1f3bf0, when it goes down to around 7-8% of my CPU (as long as vsync is on).

Is there something specific I could pursue to debug this?

**Comment by Hespian** [comment]
I'm not sure if this is related to this issue but I had to change back to compton because I noticed that picom always kept the GPU busy resulting in my notebook loosing about 1% of the battery every minute.

**Comment by Vladimir-csp** [comment]
I'm still using compton 5.1, it rarely goes above 1% when idle and only jumps to 10-20% when windows are being moved actively. Any version above that constantly uses 10-20% CPU and jumps even more during window activity.
Using GLX backend. Tried to enable/disable shadow, vsync, damage, experimental backends, lower refresh rate, no difference.
[glxinfo.txt](https://github.com/yshui/picom/files/5160543/glxinfo.txt)

I'm considering switching to picom v8 with xrender experimental backend, it seems to perform much better than in v7 and does not produce constant CPU load.

**Comment by Mark42XLII** [comment]
Exactly same issue here, e80ff85 is by far the worst of them all in idle CPU usage for me. Only happens with glx or experimental-backend, and I can confirm it's still an issue on v8.4 and the current "next" branch (af99101). I'm running Debian Bullseye on a i5-4200U with integrated graphics. One or two monitors don't seem to make much of a difference, and v5.1 is the last release that does not have this bug for me.

**Comment by yshui** [comment]
@Mark42XLII can you get a perf report? and also a debug level log might help too.

**Comment by Mark42XLII** [comment]
Done. Trying to sum it up:
Things that don't seem to affect the degree of the issue:
- Number of monitors
- Number of windows
- Vsync
- The terminal I use: tried xterm / Alacritty

Things that somewhat alleviate the issue:
- Watching videos, scrolling on the browser, or doing any render-intensive tasks such as gaming. (Yes, it makes no sense, but while idling with 3 Alacritty windows open and one running htop, I get 6 to 7% average CPU usage. While watching a video on YouTube with Firefox, around 2 to 3%. While playing very laggy Minecraft, at most 4%.)

Things that make the issue worse:
- Using the experimental GLX backend

Things that make the issue not happen at all:
- Using picom v5.1 or lower
- Using any xrender backend, experimental or not.

Of course, using blur and/or shadows brings CPU time up as well, but by a negligible amount when compared with picom with the configs I used to test all of these:
```
log-level = "debug"
log-file = "/home/mark42xlii/picom.log"
backend = "glx"
vsync=true
```
In the files, I hope you can see the CPU usage goes down as soon as I switch over to firefox and start playing a test video. Then it spikes at the end, when the video ends, and starts going down again because another video started playing on autoplay. I quickly stop it and finish the test. My picom binary is also in there, if you want to check the checksum against your build.
[perfstuff.tar.gz](https://github.com/yshui/picom/files/5683303/perfstuff.tar.gz)

**Comment by yshui** [comment]
@Mark42XLII perf doesn't has a time axis, can you do a perf again with just the high CPU usage part? thanks.

**Comment by Mark42XLII** [comment]
Oh, sorry for that, I tried separating the tests now. [perf-idling.tar.gz](https://github.com/yshui/picom/files/5683362/perf-idling.tar.gz) is the one with the high CPU usage, but I'm also sending the other one, [perf-youtube.tar.gz](https://github.com/yshui/picom/files/5683363/perf-youtube.tar.gz) just in case.

**Comment by yshui** [comment]
@Mark42XLII hmm, perf-idling.tar.gz doesn't show picom use much of the CPU

**Comment by yshui** [comment]
@Mark42XLII are you using DRI2 by any chance? can you try enabling DRI3 and see if that helps?

**Comment by yshui** [comment]
one way to find out is by running `LIBGL_DEBUG=verbose glxgears`, and there will be lines like:

`libGL: Using DRI3 for screen 0` 

in the output

**Comment by Mark42XLII** [comment]
@yshui It works! Enabled DRI3, and it totally 235% fixed it! Even with dual kawase blur the performance is amazing. Thank you so much! 

Edit: something I've noticed is that the CPU usage on the experimental GLX backend is about double the usual. By the way, this happened even while I was still using DRI2, but I forgot to mention it.

If I disable blur effects completely, I see ~11% usage when moving windows around with GLX, and 23% when using GLX with `--experimental-backends`.

About the same happens if I enable dual kawase blur and use ibhagwan's fork (since it allows me to use that blurring effect on all backends). I know it's something entirely separate from this repo, but the issue got carried over there as well, at least with my hardware.

I'll try uploading some perf data as soon as I can.

Still, I'm already happy enough that there's a solution for the idle CPU usage problem.

**Comment by Vladimir-csp** [comment]
I can confirm that enabling DRI3 fixes CPU issue!
In my case I removed `LIBGL_DRI3_DISABLE=1` from env (modesetting driver on Intel), which I added a long time ago to fix some problems I no longer remember.

**Comment by Diaoul** [comment]
I have the same issue using nouveau driver for NVIDIA card on a desktop computer. `LIBGL_DEBUG=verbose glxgears` shows `libGL: Using DRI2 for screen 0`.

Here is the rough CPU usage from htop:
* Legacy xrender 0%
* Legacy glx 5%
* Experimental xrender 1%
* Experimental glx 62%

How do I enable DRI3? Is it even possible? If possible I'd prefer to stick to the open-source driver but I'll give the proprietary one a shot to see if the issue goes away.

In any cases, if DRI3 is a requirement, would be great to have picom issue a warning when on DRI2.

**Reply by yshui** [reply]
> would be great to have picom issue a warning when on DRI2.

This is not something picom can do. Besides X server configuration, mesa is also free to choose which DRI version to use. That information is not exposed to the downstream application.

**Comment by Grafikart** [comment]
I have a 970 GTX and i'm experiencing the same issue (with nvidia proprietary driver, NVIDIA-SMI 455.45.01 / Driver Version: 455.45.01 / CUDA Version: 11.1). And when I try to run glxgears I get 

```
libGL: screen 0 does not appear to be DRI2 capable
libGL: MESA-LOADER: dlopen(/usr/lib/dri/swrast_dri.so)
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/j/.drirc: No such file or directory.
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/j/.drirc: No such file or directory.
```

I started noticing this problem quite recently with arch linux so I guess it's either a picom update or nvidia update that cause the issue (If you need more information to help debug do not hesistate).

In the meantime I disabled hardware acceleration on chrome and stopped picom and the CPU stay around 5% while watching a video (25% if picom is started with glx render).

**Comment by yshui** [comment]
@Grafikart you don't have a graphics driver installed and is using software rendering, no doubt your CPU usage is high.

**Comment by Grafikart** [comment]
@yshui I have nvidia installed and configured on xorg

```
$: lspci -nnk | grep -i vga -A3
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GM204 [GeForce GTX 970] [10de:13c2] (rev a1)
	Subsystem: ASUSTeK Computer Inc. Device [1043:8508]
	Kernel driver in use: nvidia
	Kernel modules: nouveau, nvidia_drm, nvidia
```

**Comment by yshui** [comment]
@Grafikart `lspci` shows your hardware. your driver isn't properly installed.

**Comment by Grafikart** [comment]
@yshui sorry you were totally right (nvidia inserted his configuration twice and xorg nvidia.conf was errored and ignored). So my problem was not related to this issue. Sorry :(

**Comment by Diaoul** [comment]
I've tried to force DRI3 with nouveau driver and had some screen tearing. I've switched to nvidia driver and everything is back to normal. I have glx backend with low CPU usage (1%).

I guess I'll stick to the proprietary drivers.

**Comment by yshui** [comment]
I converted this into a discussion, seems more appropriate.

@Diaoul yes, DRI3 isn't supported very well by nouveau. it's unfortunate but you'd have to use the proprietary driver

**Comment by mhykoI** [comment]
It's true that Kernel loads are decreasing. But it needs to be a little more optimized.

**Comment by hellmind** [comment]
i'm using nvidia proprietary drive and I notice cpu usage up to 10%.

**Reply by wh75er** [reply]
Same poor performance on nvidia driver

I'm getting 10%+ up to 30% when resizing

When static it's about 0.7%

Is there any options which could possibly help?

**Comment by HarshalRathore** [comment]
Hello there!
I too am having the same problem. After a little bit of backtracking to all the configurations I did related to system-graphics I found that picom started heavy CPU usage when I forced the system to use DRI2 in place of default DRI3 why you ask see this [arch wiki](https://wiki.archlinux.org/title/intel_graphics#DRI3_issues) article. I have tried switching to LTS kernel not sure if that would solve my problem and it didn't. But I think the CPU usage of picom is still a bit heavy even with DRI3 I used htop to monitor the CPU usage without picom it was 9-15% (with only browser opened with 4 tabs) with picom enabled with d-menu(with DRI3) it reached somewhat 27%+ while with DRI2 it was above 40-45%. I hope my review was helpful and sorry for any mistake it is my first time commenting on github.
I use the EndevourOS i3wm edition.
CPU is i3 5th genration.
